src/libraries/Common/src/Interop/Unix/System.Native/Interop.UnixFileSystemTypes.statvfs.cs

```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_GetFileSystemTypeAsString")]
+        private static extern string GetFileSystemTypeAsString(SafeFileHandle fd);
+
+        internal static bool TryGetFileSystemType(SafeFileHandle fd, out UnixFileSystemTypes fileSystemType) =>
+            Enum.TryParse<UnixFileSystemTypes>(GetFileSystemTypeAsString(fd), out fileSystemType);
```

Comment:
Are the names we have as the enum names all correct for this to work?

Reply:
I tested it in OpenIndiana 2021.04 VM which has the following setup:

```sh
$ df -h
Filesystem             Size   Used  Available Capacity  Mounted on
rpool/ROOT/openindiana-2021:07:18
                     47.95G  5.03G     26.76G    16%    /
/devices                  0      0          0     0%    /devices
/dev                      0      0          0     0%    /dev
ctfs                      0      0          0     0%    /system/contract
proc                      0      0          0     0%    /proc
mnttab                    0      0          0     0%    /etc/mnttab
swap                  3.58G   864K      3.58G     1%    /etc/svc/volatile
objfs                     0      0          0     0%    /system/object
bootfs                    0      0          0     0%    /system/boot
sharefs                   0      0          0     0%    /etc/dfs/sharetab
/usr/lib/libc/libc_hwcap1.so.1
                     31.79G  5.03G     26.76G    16%    /lib/libc.so.1
fd                        0      0          0     0%    /dev/fd
rpool/ROOT/openindiana-2021:07:18/var
                     47.95G  2.19G     26.76G     8%    /var
swap                  3.58G     4K      3.58G     1%    /tmp
swap                  3.58G    48K      3.58G     1%    /var/run
rpool/export         47.95G 24.50K     26.76G     1%    /export
rpool                47.95G    33K     26.76G     1%    /rpool
rpool/export/home    47.95G 24.50K     26.76G     1%    /export/home
rpool/export/home/vagrant
                     47.95G  1.08G     26.76G     4%    /export/home/vagrant
vagrant              931.55G 213.34G    718.21G    23%    /vagrant
projects             931.55G 213.34G    718.21G    23%    /projects
```

with the following C program:

```c
#include <stdio.h>
#include <errno.h>
#include <fcntl.h>
#include <inttypes.h>
#include <sys/vfs.h>

char*
// int64_t
SystemNative_GetFileSystemType(intptr_t fd)
{
    int statfsRes;
    // struct statfs statfsArgs;
    struct statvfs statfsArgs;
    // while ((statfsRes = fstatfs(fd, &statfsArgs)) == -1 && errno == EINTR) ;
    while ((statfsRes = fstatvfs(fd, &statfsArgs)) == -1 && errno == EINTR) ;
    // return statfsRes == -1 ? (int64_t)-1 : (int64_t)statfsArgs.f_type;
    return statfsRes == -1 ? "-1" : statfsArgs.f_basetype;
}

void PrintStats(const char* name) {
    int fd = open(name, O_RDONLY);

    printf("File: %s, BaseType: %s\n",
        name,
        SystemNative_GetFileSystemType((intptr_t)fd));
}

int main() {
    PrintStats("/proc");
    PrintStats("/");
    PrintStats("/tmp");
    PrintStats("/system/contract");
    PrintStats("/devices");
    PrintStats("/dev");
    PrintStats("/etc/mntab");
    PrintStats("/etc/svc/volatile");
    PrintStats("/system/object");
    PrintStats("/system/boot");
    PrintStats("/etc/dfs/sharetab");
    PrintStats("/lib/libc.so.1");
    PrintStats("/dev/fd");
    PrintStats("/var");
    PrintStats("/var/run");
    PrintStats("/rpool");
    PrintStats("/export");
    PrintStats("/export/home");
    PrintStats("/export/home/vagrant");
    PrintStats("/vagrant");

    return 0;
}
```

output of the program suggests that with the exception of procfs, all base types are reported correctly and we have them in `enum UnixFileSystemTypes`:

```
File: /proc, BaseType: 
File: /, BaseType: zfs
File: /tmp, BaseType: tmpfs
File: /system/contract, BaseType: ctfs
File: /devices, BaseType: devfs
File: /dev, BaseType: dev
File: /etc/mntab, BaseType: -1
File: /etc/svc/volatile, BaseType: tmpfs
File: /system/object, BaseType: objfs
File: /system/boot, BaseType: bootfs
File: /etc/dfs/sharetab, BaseType: sharefs
File: /lib/libc.so.1, BaseType: zfs
File: /dev/fd, BaseType: fd
File: /var, BaseType: zfs
File: /var/run, BaseType: tmpfs
File: /rpool, BaseType: zfs
File: /export, BaseType: zfs
File: /export/home, BaseType: zfs
File: /export/home/vagrant, BaseType: zfs
File: /vagrant, BaseType: sffs
```

Perhaps @jclulow might know the reason why `statvfs.f_basetype` is empty in procfs case.

Reply:
Good observation @jclulow, and sorry for the nuisance. I actually have `strdup()` around `statfsArgs.f_basetype` in the actual PR change, but missed it in the test example. `/proc` indeed returns `procfs`. However, `/etc/mntab` case is empty (which seems intentional?).

I couldn't find magic numbers for:

```
ctfs
dev
objfs
sffs
sharefs
```

in Linux or FreeBSD sources, so I haven't added them.  At usage side, the library code will consider them "unknown filesystems". If you know where could I find magic number for these, please let me know. Otherwise, I think "unknown" is file and as these filesystems are not too critical in the current usage context (it is mostly for completeness).

Reply:
> output of the program suggests that with the exception of procfs, all base types are reported correctly and we have them in `enum UnixFileSystemTypes`:

[self correction] this was wrong, as we do not have these:

```
ctfs
dev
objfs
sffs
sharefs
```

I added some of these from SmartOS in #34867, but seems like OpenIndiana has more variety.

Reply:
Ok, I think these were dropped during the refactoring. I have found some missing magic numbers and set rest to tmpfs magic as before. For Linux/macOS/FreeBSD, it will use magic and for SunOS-derivates, it will match the names.